Ast_pattern: add patterns drop and as__ (#310)

Kakadu opened the pull request.

pitag-ha commented on src/ast_pattern.mli:
```diff
@@ -109,6 +109,18 @@ val __' : ('a, 'a Loc.t -> 'b, 'b) t
+val as__ : ('a, 'b, 'c) t -> ('a, 'a -> 'b, 'c) t
+(** As-pattern. Passed the current AST to success continuation.
+
+    Pitfall. The part of AST is being passed before checking that it fits the
+    pattern. It means that success continuation may fire earlier than we realize
+    that pattern argument of [as__] doesn't fit. *)
```
Good observation! That shouldn't have a strong impact on the user though, should it? I mean, even if the success continuation set some global reference or similar, the program would stop right afterward. (Also, I'm wondering if the same happens in other combinators as well or if this is the only situation where that happens)

Kakadu replied:
> That shouldn't have a strong impact on the user though, should it? 

I had a bug about it, that took too much time to debug. 

> (Also, I'm wondering if the same happens in other combinators as well or if this is the only situation where that happens)

It looks like it is not being introduced by. I just was very disappointed about this pitfall, that's why this PR looks like it is.

pitag-ha replied:
> I had a bug about it, that took too much time to debug.

That's annoying. Would it take long to explain the bug?

Kakadu replied:
I forgot to add `()`. I wrote like in `test.ml:62` but it was required to write like on line `test.ml:66`.

pitag-ha replied:
Out of curiosity: What's your use case of `Ast_pattern`? I'm wondering because you seem to use the `parse` function yourself on the pattern as opposed to passing your pattern to `Extension.declare` or `Deriving.Args.arg` or similar.

pitag-ha replied:
Ok, sounds interesting! And I can get an idea of why you're parsing your ast_pattern yourself.

pitag-ha commented on test/patterns_as_and_drop/test.ml:
```diff
@@ -0,0 +1,69 @@
+    ~on_error:(fun () -> print_endline "error")
+    (fun _zero_expr -> Format.printf "Success and error after that\n%!");
+  printf "Correct!\n%!";
+  Ast_pattern.parse (pat ()) loc ast
+    ~on_error:(fun () () -> print_endline "error")
+    (fun _zero_expr () -> Format.printf "Success and error after that\n%!")
+    ()
```
Am I understanding correctly that here the context is a function from `unit` to `string`?

Kakadu replied:
I don't understand a question. Ast_pattern0 defines type `context ` which is isomorphic to integer, so it is not a function.

pitag-ha replied:
You're right, I need to explain myself better since in ppxlib there are lots of different concepts that we call context. What I was referring to is the type parameter `'c` in `('a, 'b, 'c) Ast_pattern.t` (i.e. the return type of the continuation `'b`). Usually, that type parameter stands for the context inside the AST in which you want to inject the derived/expanded code.

Kakadu replied:
Context is `unit -> unit` because both `() -> print_endline "error"` and `fun () -> Format.printf "Success and error after that\n%!"` have this type

pitag-ha reviewed (commented):
Thanks for the PR @Kakadu! Adding `as__` and exposing `drop` seems like a good addition to `Ast_pattern` to me.

About the tests: They're a bit tedious to read since the test code and test result are in two different files. What would you think about making `[%ppx_expect]`-tests out of them? You can see an example of that e.g. [here](https://github.com/ocaml-ppx/ppxlib/blob/main/test/extensions_and_deriving/test.ml). What do you think?

Kakadu commented:
> What would you think about making `[%ppx_expect]`-tests out of them? You can see an example of that e.g. [here](https://github.com/ocaml-ppx/ppxlib/blob/main/test/extensions_and_deriving/test.ml). What do you think?

I copied your example, and your expect-like executable writes empty file when something is wrong with the test. For example, I changed the output, got empty '.corrected' file, changed back, and got again empty '.corrected' file.  So, it is almost impossible to understand what is wrong. 

Using a clever guess I realized that it should be somehow connected to autoformatting. Indeed, when I added `[@@@ocamlformat "disable"]` to `test/extensions_and_deriving/test.ml` I managed to brake test and revert it back. But I can't do the same trick with my test, because it is already formatted wrong (in a style that your tool `test/extensions_and_deriving/.bin/expect-test` doesn't understand).

> What do you think?

Could we somehow agree on that PR without me touching an abomination called `expect-test`?

pitag-ha commented:
> For example, I changed the output, got empty '.corrected' file, changed back, and got again empty '.corrected' file. So, it is almost impossible to understand what is wrong.

Yes, I agree, that's very annoying. That happens when you don't have any `%%expect` extension node in your test file (for example, if you have a typo like `%%espect` or `%expect` or so). That's because we're using our own expect test version as opposed to the ppx_expect opam package to avoid circular dependencies.

> it should be somehow connected to autoformatting.

The reason why we disable ocamlformat for expect test files is that our expect test doesn't respect ocamlformat formatting, so promoting `dune runtest` and `dune build @fmt` would yield conflicting results when ocamlformat isn't disabled. 

In case you want to give it another try: I'd expect the following to work (you'd simply have to `dune promote` it):
```
(* The `as__` combinator captures the whole pattern it is given
  in addition to the internally captured values in that pattern.
  The value it captures is passed to the continuation before the internally captured values. *)
let parse_result =
  let loc = Location.none in
  let ast = [%expr List.length xs = 0] in
  let pat =
    let open Ast_pattern in
    let length () =
      as__
        (pexp_apply
           (pexp_ident (ldot (lident (string "List")) (string "length")))
           ((nolabel ** drop) ^:: nil))
    in
    let zero () = as__ (pexp_constant __) in

    pexp_apply
      (pexp_ident (lident (string "=")))
      ((nolabel ** length ()) ^:: (nolabel ** zero ()) ^:: nil)
  in

  Ast_pattern.parse pat loc ast
    (fun first_as_captured_val second_as_captured_val intrenally_captured_val ->
      (first_as_captured_val, second_as_captured_val, intrenally_captured_val))
[%%expect{|
|}]
```
It's pretty much your second test with one modification: it tests the behavior of `as__` itself. Testing the pitfall as well is also great though!

If you still really don't want to use the expect test framework, we can find another solution.

pitag-ha commented on test/patterns_as_and_drop/test.ml:
```diff
@@ -0,0 +1,81 @@
+    (fun _zero_expr () -> Format.printf "Success and error after that\n%!")
+    ()
+
+(* But this pitfall is not introduced by [as__], it was before too. *)
+let () =
+  let ast = [%expr string_of_int 42] in
+  let pat =
+    let open Ast_pattern in
+    pexp_apply (pexp_ident (lident __)) ((nolabel ** eint (int 42)) ^:: nil)
+  in
+
+  Ast_pattern.parse pat loc ast
+    ~on_error:(fun () -> print_endline "error")
+    (fun s -> printf "success %s, but actually not" s)
```
How does this new test reveal the pitfall? At first sight, it seems like the `ast` you define simply fulfils the pattern you define. So why is this "actually no success"? (I'm probably be missing something)

Kakadu replied:
Actually, it was a little bit crash and this test doesn't represent what I wanted. Not it should be better.

pitag-ha replied:
Thanks!

Kakadu commented:
> If you still really don't want to use the expect test framework, we can find another solution.

It looks like I managed to rewrite with expect-tests

pitag-ha reviewed (commented):
Thanks for the improvements! I only have a couple of minor comments about the documentation that I'd like to be addressed before merging (notice that my suggestions aren't formatted).

Kakadu commented:
@pitag-ha  ^^

pitag-ha reviewed (approved):
Thanks for the change. LGTM!
